lib/renderer.js

```diff
@@ -83,17 +89,24 @@ class Renderer extends MarkedRenderer {
 
   // Prepend root to image path
   image(href, title, text) {
-    const { options } = this;
-
-    if (!parse(href).hostname && !options.config.relative_link
-      && options.prependRoot) {
-      href = url_for.call(options, href);
+    const { hexo, options } = this;
+    const { relative_link } = hexo.config;
+    const { lazyload, prependRoot, postId } = options;
+
+    if (!/^(#|\/\/|http(s)?:)/.test(href) && !relative_link && prependRoot) {
+      if (!/^(\/|\\)/.test(href) && postId) {
+        const PostAsset = hexo.model('PostAsset');
+        // slug requires platform-specific path
+        const asset = PostAsset.findOne({ post: postId, slug: href.replace(/\/|\\/g, sep) });
```

Comment:
Instead of using `findOne`, is it possible to use `findById` directly?

https://github.com/hexojs/hexo/blob/master/lib/plugins/processor/post.js

Also, each post should only find PostAsset for once. A cache should be introduced.

Reply:
I am suggesting `WeakMap`: https://mdn.io/WeakMap

`hexojs/hexo` already utilise `WeakMap` as Cache:

https://github.com/hexojs/hexo/blob/9b41e6d30bef0418d312e783c9bba6d290c1e4fb/lib/hexo/index.js#L34
https://github.com/hexojs/hexo/blob/9b41e6d30bef0418d312e783c9bba6d290c1e4fb/lib/hexo/index.js#L56
https://github.com/hexojs/hexo/blob/9b41e6d30bef0418d312e783c9bba6d290c1e4fb/lib/hexo/index.js#L72

Since `post` here is an object provided by `warehouse`, `WeakMap` could fit our need.

Reply:
```js
const { Cache } = require('hexo-util');
const postAssetCache = new Cache();

// ...

const asset = postAssetCache.apply(postId, () => PostAsset.findOne({ post: postId, slug: href.replace(/\/|\\/g, sep) }));
```

And see if there is any performance improvement.